batchflow/opensets/coco.py

```diff
@@ -0,0 +1,112 @@
+
+logger = logging.getLogger('COCO')
+
+
+class BaseCOCO(ImagesOpenset):
+    """ Base class for COCO datasets. """
+
+    TRAIN_IMAGES_URL = 'http://images.cocodataset.org/zips/train2017.zip'
+    TEST_IMAGES_URL = 'http://images.cocodataset.org/zips/val2017.zip'
+
+    def __init__(self, *args, preloaded=None, **kwargs):
+        super().__init__(*args, preloaded=preloaded, **kwargs)
+
+
+    @parallel(init='_get_from_urls', post='_post_fn', target='t')
+    def download(self, url, folder, train_val, path=None):
+        """ Download the archives and extracts it's content. Downloading is performed in parallel manner.
+        Set of URL's to download from is defined in the `_get_from_urls` method.
+        The aggregation of the content from all archives is performed in `_post_fn` method.
+        """
+        logger.info('Downloading %s', url)
+        if path is None:
+            path = tempfile.gettempdir()
+        filename = basename(url)
+        localname = os.path.join(path, filename)
+        if not os.path.isfile(localname):
+            r = requests.get(url, stream=True)
+            file_size = int(r.headers['Content-Length'])
+            chunk = 1
+            chunk_size = 1024 * 1000 #MBs
+            num_bars = int(file_size / chunk_size)
+            with open(localname, 'wb') as f:
+                for chunk in tqdm.tqdm(r.iter_content(chunk_size=chunk_size), total=num_bars,
+                                       unit='MB', desc=filename, leave=True):
+                    f.write(chunk)
+        return self._extract_if_not_exist(localname, folder, train_val)
+
+
+class COCOSegmentation(BaseCOCO):
+    """ The dataset for COCOStuff challenge for pixel-wise segmentation. Total size 18GB.
+    Contains 118060 train and 4990 test images/masks.
+
+    Notes
+    -----
+    - Datasets contain both grayscale and colored images
+      Argument `drop_grayscale` controls whether grayscale images should be dropped.
+    """
+
+    MASKS_URL = 'http://calvin.inf.ed.ac.uk/wp-content/uploads/data/cocostuffdataset/stuffthingmaps_trainval2017.zip'
+
+    def __init__(self, *args, drop_grayscale=True, **kwargs):
+        self.drop_grayscale = drop_grayscale
+        super().__init__(*args, **kwargs)
+
+
+    @property
+    def _get_from_urls(self):
+        """ List of URL to download from, folder where to extract, and indicator whether its train or val part. """
+        iterator = zip([self.TRAIN_IMAGES_URL, self.TEST_IMAGES_URL, self.MASKS_URL, self.MASKS_URL],
+                       ['COCOImages', 'COCOImages', 'COCOMasks', 'COCOMasks'],
+                       ['train2017', 'val2017', 'train2017', 'val2017'])
+        return [[url, folder, train_val] for url, folder, train_val in iterator]
+
+    def _extract_archive(self, localname, extract_to):
+        with ZipFile(localname, 'r') as archive:
+            archive.extractall(extract_to)
+
+    def _extract_if_not_exist(self, localname, folder, train_val):
```

Comment:
Do you really need this function? Looks like it would be exactly the same for any other `CoCo` datasets (if we were to add more), and it would be way more clear if the contents of that function were part of the `download` method of `BaseCOCO` class
And you won't need that strange return of path from an extracting function